### Import Modules

In [1]:
import re
import pickle
import pandas as pd
import string

### Setting up Dataframe

In [2]:
# opening pickle file
with open('comedian.pickle', 'rb') as f:
    heading = pickle.load(f)

In [3]:
heading[:5]

['BILLY CONNOLLY: HIGH HORSE TOUR LIVE (2016) – FULL TRANSCRIPT',
 'HANNAH GADSBY: DOUGLAS (2020) – FULL TRANSCRIPT',
 'PATTON OSWALT: I LOVE EVERYTHING (2020) – FULL TRANSCRIPT',
 'RUSSELL PETERS: DEPORTED (2020) – FULL TRANSCRIPT',
 'JIMMY O. YANG: GOOD DEAL (2020) – FULL TRANSCRIPT']

In [4]:
# make a datafram using the heading
comedian_df = pd.DataFrame({'Heading': heading})

In [5]:
comedian_df.head()

Heading
0  BILLY CONNOLLY: HIGH HORSE TOUR LIVE (2016) – ...
1    HANNAH GADSBY: DOUGLAS (2020) – FULL TRANSCRIPT
2  PATTON OSWALT: I LOVE EVERYTHING (2020) – FULL...
3  RUSSELL PETERS: DEPORTED (2020) – FULL TRANSCRIPT
4  JIMMY O. YANG: GOOD DEAL (2020) – FULL TRANSCRIPT

### Splitting Names

In most cases the format in the heading column is "firstname lastname:Title (Year)". This makes it easy to get most comedian names.

In [6]:
# Some headings don't have a colon. This is a list of words to split comedian names for most other cases.

split_words =[':', ' On ', ' At ', ' About ', ' Live', 'Stand-Up', '–', '…', 'Award-Winning', 'Monologue', ' & ', 'Matters', 'E Il', 'Tries']

In [7]:
# loop through each item in the heading, splitting the text using a loop on each split word
nm=[]
for name in comedian_df['Heading']:
        for word in split_words:
            name = (name.split(word)[0].title().rstrip())
        nm.append(name)

comedian_df['Name'] = nm

In [8]:
comedian_df.head()

Heading            Name
0  BILLY CONNOLLY: HIGH HORSE TOUR LIVE (2016) – ...  Billy Connolly
1    HANNAH GADSBY: DOUGLAS (2020) – FULL TRANSCRIPT   Hannah Gadsby
2  PATTON OSWALT: I LOVE EVERYTHING (2020) – FULL...   Patton Oswalt
3  RUSSELL PETERS: DEPORTED (2020) – FULL TRANSCRIPT  Russell Peters
4  JIMMY O. YANG: GOOD DEAL (2020) – FULL TRANSCRIPT   Jimmy O. Yang

In [9]:
# testing another example of the dataframe. Note - the name in line 50 needs work

comedian_df.iloc[49:51]

Heading  \
49  DAVE CHAPPELLE: STICKS AND STONES | EPILOGUE: ...   
50  Brazil, Corruption and the Amazon Rainforest |...   

                                                 Name  
49                                     Dave Chappelle  
50  Brazil, Corruption And The Amazon Rainforest |...

#### Fixing Names

In [10]:
# Most names look correct, but we have to make an adjustment on a few

name_index = [11, 50, 57, 58, 63, 83, 86, 88, 100, 101, 103, 112, 148, 224, 231, 235, 237, 338]

In [11]:
# List of corrected comedian names
name_fix = ['Louis C.K.',
'Hasan Minhaj',
'Katherine Ryan',
'Katherine Ryan',
'Gina Yashere',
'George Carlin',
'Gabriel Iglesias',
'John Leguizamo',
'George Carlin',
'Tom Segura',
'Joe Mande',
'Dave Chappelle',
'Patton Oswalt',
'Richard Pryor',
'Dave Attell',
'Daniel Tosh',
'George Carlin',
'Pauline Kael',
]


#### Looping through corrected names

In [12]:
# Loop through adjustments using the 2 lists above
for i, n in zip(name_index, name_fix):
        comedian_df.Name.iloc[i] = n

In [13]:
# a look at the revised dataframe
comedian_df.iloc[49:51]

Heading            Name
49  DAVE CHAPPELLE: STICKS AND STONES | EPILOGUE: ...  Dave Chappelle
50  Brazil, Corruption and the Amazon Rainforest |...    Hasan Minhaj

### Splitting Year

The heading usually has information about the year of each stand-up set. Follow a similar process to split this out

In [14]:
# loop uses regex to find 4 digit numbers as year. Then add Year to dataframe
yr = []    
for name in comedian_df['Heading']:
    try:
        yr.append(re.search(r'\b\d{4}\b',name).group(0))
    except:
        yr.append('missing year')
comedian_df['Year'] = yr

In [15]:
# those without 4 digits numbers are classified as missing year. This is a list of them all
missing_yr = comedian_df['Year'] == 'missing year'

In [16]:
# viewing the dataframe again
comedian_df.iloc[49:52]

Heading            Name  \
49  DAVE CHAPPELLE: STICKS AND STONES | EPILOGUE: ...  Dave Chappelle   
50  Brazil, Corruption and the Amazon Rainforest |...    Hasan Minhaj   
51  DAVE CHAPPELLE: STICKS & STONES (2019) – FULL ...  Dave Chappelle   

            Year  
49  missing year  
50  missing year  
51          2019

In [17]:
comedian_df.head()

Heading            Name  Year
0  BILLY CONNOLLY: HIGH HORSE TOUR LIVE (2016) – ...  Billy Connolly  2016
1    HANNAH GADSBY: DOUGLAS (2020) – FULL TRANSCRIPT   Hannah Gadsby  2020
2  PATTON OSWALT: I LOVE EVERYTHING (2020) – FULL...   Patton Oswalt  2020
3  RUSSELL PETERS: DEPORTED (2020) – FULL TRANSCRIPT  Russell Peters  2020
4  JIMMY O. YANG: GOOD DEAL (2020) – FULL TRANSCRIPT   Jimmy O. Yang  2020

#### Fixing Years

In [18]:
# creating an index for the missing year list
missing_yr_index = [i for i, val in enumerate(missing_yr) if val]

In [19]:
missing_yr_index

[16,
 19,
 22,
 49,
 50,
 54,
 65,
 83,
 101,
 107,
 108,
 142,
 223,
 236,
 238,
 311,
 328,
 338]

In [20]:
# had to manually look up some of the missing years

yr_list = ['1993',
'1965',
'2015',
'2019',
'2019',
'2019',
'2004',
'1997',
'2011',
'2012',
'2014',
'2007',
'1996',
'1999',
'2014',
'1972',
'2014',
'1980']

#### Looping through missing years

In [21]:
# appling the missing years
for i, y in zip(missing_yr_index, yr_list):
        comedian_df.Year.iloc[i] = y

In [22]:
# checking the df - years have been added
comedian_df.iloc[49:51]

Heading            Name  Year
49  DAVE CHAPPELLE: STICKS AND STONES | EPILOGUE: ...  Dave Chappelle  2019
50  Brazil, Corruption and the Amazon Rainforest |...    Hasan Minhaj  2019

In [23]:
# top 20 years by count, most are recent
comedian_df.Year.value_counts()[:20]

2018    43
2017    40
2019    40
2016    27
2015    22
2014    16
2013    15
2020    14
2012    14
2010    13
2011    11
2007     9
1999     8
2006     8
2008     6
2004     6
2009     5
1992     4
1996     4
1983     3
Name: Year, dtype: int64

### Splitting Title

We can also split the title of each stand-up set from the heading

In [24]:
# Start by splitting on hyphen or parentheses or brackets
split_titles =['–', ' (', ' [']

In [25]:
# Note this time we apply 'capwords' rather than 'title' as capwords works better with apostrophes in titles
# The second part of the loop splits on colon, taking the second part of the split. 
# We need a try / except for cases without a colon

ti=[]
for ttl in comedian_df['Heading']:
        for word in split_titles:
            ttl = string.capwords(ttl.split(word)[0]).rstrip()
    
        try:
            ttl = (ttl.split(":")[1].lstrip())
        except:
            ttl = ttl
        ti.append(ttl)

comedian_df['Title'] = ti


In [26]:
# viewing the dataframe, note the dots for an extended title on line 50
comedian_df.iloc[49:51]

Heading            Name  Year  \
49  DAVE CHAPPELLE: STICKS AND STONES | EPILOGUE: ...  Dave Chappelle  2019   
50  Brazil, Corruption and the Amazon Rainforest |...    Hasan Minhaj  2019   

                                                Title  
49                       Sticks And Stones | Epilogue  
50  Brazil, Corruption And The Amazon Rainforest |...

#### Fixing Titles

In [27]:
# a few titles need refinement.

title_index = [10,
50,
57,
58,
63,
65,
74,
88,
101,
103,
142,
148,
156,
180,
223,
224,
229,
231,
235,
251,
305,
324,
328,
338]

In [28]:
title_fix = ['The Late Show With David Letterman',
'Brazil, Corruption And The Amazon Rainforest',
'In Trouble',
'Glitter Room',
'The Standups',
'Babies And Abortion',
'Live On Broadway',
'Latin History For Morons',
'Comedy Central Presents',
'Award-Winning Comedy Special',
'Nationalism',
'Comedy Central Presents',
'Glorious',
'Live at Laff Stop',
'Abortion and “Pro-Lifers”',
'Monologue at Saturday Night Live',
'Tries Stand-up For The First Time',
'Comedy Central Presents',
'Comedy Central Presents',
'Here And Now',
'2017 White House Correspondents’ Dinner',
'2011 White House Correspondents’ Dinner',
'Gun Control',
'The Shining Devolution']

#### Looping through title fix

In [29]:
for i, t in zip(title_index, title_fix):
        comedian_df.Title.iloc[i] = t

In [30]:
# viewing same titles after the fix has been applied
comedian_df.iloc[49:51]

Heading            Name  Year  \
49  DAVE CHAPPELLE: STICKS AND STONES | EPILOGUE: ...  Dave Chappelle  2019   
50  Brazil, Corruption and the Amazon Rainforest |...    Hasan Minhaj  2019   

                                           Title  
49                  Sticks And Stones | Epilogue  
50  Brazil, Corruption And The Amazon Rainforest

### Add Link to Dataframe

Now we can add the link and the transcript to our dataframe

In [31]:
# opening pickle file
with open('link.pickle', 'rb') as f:
    link = pickle.load(f)

In [32]:
comedian_df['Link'] = link

In [33]:
comedian_df

Heading            Name  Year  \
0    BILLY CONNOLLY: HIGH HORSE TOUR LIVE (2016) – ...  Billy Connolly  2016   
1      HANNAH GADSBY: DOUGLAS (2020) – FULL TRANSCRIPT   Hannah Gadsby  2020   
2    PATTON OSWALT: I LOVE EVERYTHING (2020) – FULL...   Patton Oswalt  2020   
3    RUSSELL PETERS: DEPORTED (2020) – FULL TRANSCRIPT  Russell Peters  2020   
4    JIMMY O. YANG: GOOD DEAL (2020) – FULL TRANSCRIPT   Jimmy O. Yang  2020   
..                                                 ...             ...   ...   
338   THE SHINING: DEVOLUTION – REVIEW BY PAULINE KAEL    Pauline Kael  1980   
339  DAVE CHAPPELLE: STICKS & STONES (2019) – FULL ...  Dave Chappelle  2019   
340  KEVIN HART: SERIOUSLY FUNNY (2010) – FULL TRAN...      Kevin Hart  2010   
341   TOM SEGURA: DISGRACEFUL (2018) – FULL TRANSCRIPT      Tom Segura  2018   
342    HANNAH GADSBY: NANETTE (2018) – FULL TRANSCRIPT   Hannah Gadsby  2018   

                      Title                                               Link  
0      High Horse Tour Live  https://scrapsfromtheloft.com/2020/05/27/billy...  
1                   Douglas  https://scrapsfromtheloft.com/2020/05/26/hanna...  
2         I Love Everything  https://scrapsfromtheloft.com/2020/05/20/patto...  
3                  Deported  https://scrapsfromtheloft.com/2020/05/10/russe...  
4                 Good Deal  https://scrapsfromtheloft.com/2020/05/10/jimmy...  
..                      ...                                                ...  
338  The Shining Devolution  https://scrapsfromtheloft.com/2016/10/16/shini...  
339         Sticks & Stones  https://scrapsfromtheloft.com/2019/08/26/dave-...  
340         Seriously Funny  https://scrapsfromtheloft.com/2017/09/02/kevin...  
341             Disgraceful  https://scrapsfromtheloft.com/2018/01/13/tom-s...  
342                 Nanette  https://scrapsfromtheloft.com/2018/07/21/hanna...  

[343 rows x 5 columns]

### Add Transcript to Dataframe

In [34]:
# opening pickle file
with open('transcript2.pickle', 'rb') as f:
    transcript = pickle.load(f)

In [35]:
comedian_df['Transcript'] = transcript

comedian_df.Transcript[342][:500]

#### Adding the length of each transcript to the dataframe

In [38]:
# loop counts the number of characters in each transcript
length =[]
for t in comedian_df['Transcript']:
    lngth = len(t)
    length.append(lngth)

comedian_df['Length'] = length

In [39]:
# view first example
len(comedian_df['Transcript'][0])

65834

#### Adding word count of each transcript to the dataframe

In [40]:
# loop counts the number of words in each transcript

In [41]:
words = []
for w in comedian_df['Transcript']:
    count = len(w.split())
    words.append(count)
    
comedian_df['WordCount'] = words

In [42]:
# viewing comedian names. Making sure that there are no alternative spellings - Dave vs David for example
sorted(comedian_df.Name.unique())

['Adam Devine',
 'Adam Sandler',
 'Adel Karam',
 'Al Madrigal',
 'Ali Wong',
 'Amanda Seales',
 'Amy Schumer',
 'Andy Woodhull',
 'Anjelah Johnson',
 'Anthony Jeselnik',
 'Ari Shaffir',
 'Arsenio Hall',
 'Aziz Ansari',
 'Bert Kreischer',
 'Big Jay Oakerson',
 'Bill Burr',
 'Bill Hicks',
 'Bill Maher',
 'Billy Connolly',
 'Bo Burnham',
 'Brad Williams',
 'Brent Morin',
 'Brian Regan',
 'Bridget Everett',
 'Cedric The Entertainer',
 'Chelsea Peretti',
 'Chris D’Elia',
 'Chris Rock',
 'Chris Tucker',
 'Christina Pazsitzky',
 'Colin Quinn',
 'Craig Ferguson',
 'Cristela Alonzo',
 'D.L. Hughley',
 'Dan Soder',
 'Dana Carvey',
 'Daniel Sloss',
 'Daniel Tosh',
 'Dave Allen',
 'Dave Attell',
 'Dave Chappelle',
 'David Cross',
 'Demetri Martin',
 'Deray Davis',
 'Dick Gregory',
 'Donald Glover',
 'Doug Stanhope',
 'Dylan Moran',
 'Eddie Griffin',
 'Eddie Izzard',
 'Eddie Murphy',
 'Ellen Degeneres',
 'Emily Heller',
 'Enissa Amani',
 'Erik Griffin',
 'Fahim Anwar',
 'Frankie Boyle',
 'Fred Armi

In [43]:
# Top Comedians by transcript count
comedian_df.Name.value_counts()[:30]

George Carlin           24
Dave Chappelle          13
Louis C.K.              11
Ricky Gervais            8
Jim Jefferies            8
Bill Burr                8
John Mulaney             7
Kevin Hart               6
Tom Segura               6
Chris Rock               6
Doug Stanhope            5
Iliza Shlesinger         5
Trevor Noah              5
Richard Pryor            5
Eddie Izzard             4
Amy Schumer              4
Patton Oswalt            4
Daniel Tosh              4
Sarah Silverman          4
Stewart Lee              4
Jim Norton               4
Jimmy Carr               4
Bert Kreischer           4
Sebastian Maniscalco     4
Gabriel Iglesias         4
Bill Maher               3
Frankie Boyle            3
Jo Koy                   3
Kevin Bridges            3
Bill Hicks               3
Name: Name, dtype: int64

In [44]:
# backup before removing duplicates
backupcopy_df = comedian_df